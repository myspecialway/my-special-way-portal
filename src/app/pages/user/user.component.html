<div data-test-id="users-page" class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">
          {{'USERS.USERS_TITLE' | translate}}
          </span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button mat-raised-button color="primary" (click)="addNewUser()" title="{{'USERS.USERS_INSERT' | translate}}">
          <i class="material-icons">
           add
          </i> {{'USERS.USERS_INSERT' | translate}}
       
        </button>
      </div>
    </div>
    <div class="row msw-table-container">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table"
                   matSort matSortActive="firstName" matSortDisableClear matSortDirection="asc">

          <!-- ID Column -->
          <ng-container matColumnDef="name">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_NAME' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)"> {{row.firstname}} {{ row.lastname }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="class">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_CLASS' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{(row.class?.name) || '' }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_USERNAME' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{row.username}}</mat-cell>
          </ng-container>


          <!-- Name Column -->
          <ng-container matColumnDef="type">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_USERROLE' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{toHebrew(row.role)}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="email">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_EMAIL' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{row.email}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="recoverpassword">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button class="recover-password-button" mat-icon-button color="primary"> {{'USERS.USERS_RECOVERPASS' | translate}}</button>
            </mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button class="delete-button" mat-icon-button color="primary" (click)="deleteUser(row)" title="{{'COMMONS.DELETE' | translate}}"><i class="material-icons">delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;">
        </mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
