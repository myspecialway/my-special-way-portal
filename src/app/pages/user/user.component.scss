@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.recover-password-button {
  text-decoration: underline;
  display: none;
}

.mat-row:hover {
  .recover-password-button {
    display: block;
  }
}
