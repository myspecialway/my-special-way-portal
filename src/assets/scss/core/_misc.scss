body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
