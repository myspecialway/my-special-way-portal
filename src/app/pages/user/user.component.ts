import { Component, ElementRef, OnInit, ViewChild, AfterViewInit } from '@angular/core';
import { MatDialog, MatSort, MatTableDataSource } from '@angular/material';
import { UserService } from './services/user.service';
import { User, UserType } from '../../models/user.model';
import { AddUserDialogComponent } from './dialogs/add/add-user.dialog';
import { DeleteUserDialogComponent } from './dialogs/delete/delete-user.dialog';
import * as _ from 'lodash';
import { UpdateUserDialogComponent } from './dialogs/update/update-user.dialog';

@Component({
  selector: 'app-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss'],
})
export class UserComponent implements OnInit, AfterViewInit {
  displayedColumns = ['name', 'username', 'class', 'type', 'enterPersonalArea', 'editDetails', 'deleteUser'];
  dataSource = new MatTableDataSource<User>();
  resultsLength = 0;

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  constructor(private userService: UserService, public dialog: MatDialog) {}

  ngOnInit(): void {
    this.userService.getAllUsers().subscribe((data) => {
      this.dataSource.data = data;
    });
  }

  ngAfterViewInit() {
    this.dataSource.sort = this.sort;
  }
  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // Datasource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }

  toHebrew(type: UserType) {
    return UserType[type];
  }

  addNewUser() {
    const dialogRef = this.dialog.open(AddUserDialogComponent, {
      data: { user: User },
      height: '368px',
      width: '630px',
    });
    dialogRef.afterClosed().subscribe((data) => {
      if (data) {
        const newUser: User = this._createNewUser(data);
        this.userService.create(newUser);
      }
    });
  }
  deleteUser(_id: number, firstName: string, lastName: string, userType: UserType) {
    const dialogRef = this.dialog.open(DeleteUserDialogComponent, {
      data: { _id, firstName, lastName, userType },
      height: '275px',
      width: '360px',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result === true) {
        this.userService.delete(_id);
      }
    });
  }
  updateUser(_id: number, firstname: string, lastname: string, email: string, username: string, role: any, Class: any) {
    const dialogRef = this.dialog.open(UpdateUserDialogComponent, {
      data: { _id, firstname, lastname, email, username, role, Class },
      height: '376px',
      width: '631px',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result) {
        const relevantUser = _.find(this.dataSource.data, { _id });
        const tempUser = _.assign({}, relevantUser, result);

        this.userService.update(tempUser);
      }
    });
  }

  _createNewUser(userData: any): User {
    const user: User = new User();
    user.firstname = userData.firstName;
    user.lastname = userData.lastName;
    user.username = userData.userName;
    user.email = userData.email;
    user.role = userData.userType;
    user.Class = userData.Class;
    return user;
  }
}
