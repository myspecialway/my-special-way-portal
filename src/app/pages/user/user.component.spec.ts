import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRowDef, MatRowDef, MatHeaderRow, MatDialog, MatSort } from '@angular/material';
import { UserComponent } from './user.component';
import { UserService } from './services/user.service';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { TranslateCustomLoader } from '../../../mocks/translate.stub';

import {
  Overlay,
  ScrollStrategyOptions,
  ScrollDispatcher,
  ViewportRuler,
  OverlayContainer,
  OverlayPositionBuilder,
  OverlayKeyboardDispatcher,
} from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { UserType } from '../../models/user.model';
import { userTestData } from '../../../mocks/assets/users.mock';
import { Observable } from 'rxjs-compat';

describe('user component', () => {
  let userServiceMock: Partial<UserService>;
  let userDialogMock: Partial<MatDialog>;

  beforeEach(async () => {
    userServiceMock = {
      getAllUsers: jest.fn().mockReturnValueOnce(Observable.of(userTestData.users)),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };

    userDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: jest.fn().mockReturnValue(Observable.of(true)),
      }),
    };

    TestBed.configureTestingModule({
      imports: [
        TranslateModule.forRoot({
          loader: {
            provide: TranslateLoader,
            useClass: TranslateCustomLoader,
          },
        }),
      ],
      declarations: [UserComponent, MatHeaderRow, MatRowDef, MatHeaderRowDef, MatSort],
      providers: [
        { provide: MatDialog, useValue: userDialogMock },
        { provide: UserService, useValue: userServiceMock },
        Overlay,
        ScrollStrategyOptions,
        ScrollDispatcher,
        Platform,
        ViewportRuler,
        OverlayContainer,
        OverlayPositionBuilder,
        OverlayKeyboardDispatcher,
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling addNewUser function', () => {
    (userServiceMock.create as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();

    fixture.componentInstance.addNewUser();
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling deleteUser function', () => {
    (userServiceMock.delete as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });

    const user = {
      _id: 123,
      username: 'uname',
      firstname: 'sad',
      lastname: 'asd',
      role: UserType.PRINCIPLE,
      password: '123',
      email: 'asdd@sdsd.com',
    };
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();

    fixture.componentInstance.deleteUser(user);
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling updateeUser function', () => {
    (userServiceMock.update as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve(1);
    });
    const user = {
      _id: 123,
      username: 'uname',
      firstname: 'sad',
      lastname: 'asd',
      role: UserType.PRINCIPLE,
      password: '123',
      email: 'asdd@sdsd.com',
    };
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();

    fixture.componentInstance.updateUser(user);
    const DialogMock = TestBed.get(MatDialog);
    expect(DialogMock.open).toHaveBeenCalled();
  });

  it('should load users from service on page load ', () => {
    (userServiceMock.create as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve();
    });
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    const serviceMock = TestBed.get(UserService);
    expect(serviceMock.getAllUsers).toHaveBeenCalled();
  });

  it('should load correct number of users ', async () => {
    (userServiceMock.getAllUsers as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve();
    });
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(2);
  });

  it('should load zero users when recieves promise reject', async () => {
    (userServiceMock.getAllUsers as jest.Mock).mockReset();
    (userServiceMock.getAllUsers as jest.Mock).mockReturnValueOnce(Observable.of([]));
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.dataSource.data.length).toEqual(0);
  });

  it('should use the type methode to translate TEACHER to hebrew', async () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.toHebrew('TEACHER' as UserType)).toEqual('מורה');
  });
  it('should use the type methode to translate ORINCIPAL to hebrew', async () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    expect(fixture.componentInstance.toHebrew('PRINCIPLE' as UserType)).toEqual('מנהל');
  });
  it('should clean the filter before aplying to table', async () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.detectChanges();
    await fixture.whenRenderingDone();
    fixture.componentInstance.applyFilter('  AA!!BB  ');
    expect(fixture.componentInstance.dataSource.filter).toEqual('aa!!bb');
  });
});
