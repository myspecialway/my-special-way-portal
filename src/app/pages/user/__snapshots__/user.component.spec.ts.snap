// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user component should render component as described in snapshot 1`] = `
<app-user
  dataSource={[Function MatTableDataSource]}
  dialog={[Function Object]}
  displayedColumns={[Function Array]}
  resultsLength="0"
  sort={[Function MatSort]}
  subCollector={[Function Subscription]}
  table={[Function ElementRef]}
  userService={[Function Object]}
>
  <div
    class="main-content"
    data-test-id="users-page"
  >
    <div
      class="container-fluid"
    >
      <div
        class="header-menu-container msw-table-tools-header"
      >
        <div
          class="header-table-count"
        >
          <div>
            <span
              class="header-table-count-label"
            >
              משתמשים
            </span>
            <span
              class="header-table-count-number"
            >
              (2)
            </span>
          </div>
        </div>
        <div
          class="example-header msw-table-tools-header"
        >
          <button
            color="primary"
            mat-raised-button=""
          >
            <i
              class="material-icons"
            >
               add 
            </i>
             הוספת משתמש חדש 
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          <mat-table
            class="msw-table"
            matsort=""
            matsortactive="firstName"
            matsortdirection="asc"
            matsortdisableclear=""
            ng-reflect-active="firstName"
            ng-reflect-direction="asc"
            ng-reflect-disable-clear=""
          >
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            --&gt; 
            
            
            
            
          </mat-table>
        </div>
      </div>
    </div>
  </div>
</app-user>
`;
